### PR TITLE
 Provision everything on CR creation 

### PR DESCRIPTION
**Pull request by grdryn**
JIRA: https://issues.jboss.org/browse/AEROGEAR-9187

This is mostly just copypasta from the unifiedpush-operator, since from a deployment point-of-view they're very similar.

----

# Verification steps

* Checkout this branch
* As a cluster admin user in OpenShift (because of role creation), run `make cluster/prepare`. This will create an app-metrics project and other things needed for the operator.
* Deploy the operator using the image built for this PR:
```
cat deploy/operator.yaml | sed -e "s|quay.io/aerogear/app-metrics-operator:master|quay.io/aerogear/app-metrics-operator:$(git rev-parse --short HEAD|g" | kubectl create -f -
```
* Check that the operator has started successfully with like `kubectl get po` or whatever, and have a look at the logs to make sure there's no obvious errors in there.

* Create an app-metrics instance by adding a CR! :smiley_cat: :
```
kubectl apply -f deploy/crds/metrics_v1alpha1_appmetricsservice_cr.yaml
```

* Once everything has been created, you should be able to login to the service, through the oauth-proxy by going to the route in your browser. On linux, something like this should probably work to open it in your default browser:
```
xdg-open https://$(oc get route -o json | jq -r .items[0].spec.host)
```

* If you're happy with it, you can optionally remove everything again by running `make cluster/clean`.

### CODE REVIEW

File: Gopkg.lock [hunk 1]
```diff
@@ -377,6 +377,20 @@
   revision = "4b7aa43c6742a2c18fdef89dd197aaae7dac7ccd"
   version = "1.0.1"
 
+[[projects]]
```
Comment (grdryn) [inline Gopkg.lock:380]: Hmm, running `dep ensure` pulls in these dependencies once they're referenced, and it updates `Gopkg.lock` but not `Gopkg.toml`. I'm not sure if that's important.

File: pkg/controller/appmetricsservice/postgresql.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func newPostgresqlPersistentVolumeClaim(cr *metricsv1alpha1.AppMetricsService) (*corev1.PersistentVolumeClaim, error) {
+	pvcSize, err := resource.ParseQuantity("10Gi")
```
Comment (grdryn) [inline pkg/controller/appmetricsservice/postgresql.go:18]: This should probably be configurable from the CR. Thoughts?

**Comment by aliok** [comment]
:eye:

File: deploy/crds/metrics_v1alpha1_appmetricsservice_cr.yaml [hunk 3]
```diff
@@ -3,5 +3,3 @@ kind: AppMetricsService
 metadata:
   name: example-appmetricsservice
 spec:
```
Comment (aliok) [inline deploy/crds/metrics_v1alpha1_appmetricsservice_cr.yaml]: Need to delete the empty `spec`.

```
[aliok@localhost app-metrics-operator]$ kubectl apply -f deploy/crds/metrics_v1alpha1_appmetricsservice_cr.yaml
The AppMetricsService "example-appmetricsservice" is invalid: []: Invalid value: map[string]interface {}{"spec":interface {}(nil), "apiVersion":"metrics.aerogear.com/v1alpha1", "kind":"AppMetricsService", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"metrics.aerogear.com/v1alpha1\",\"kind\":\"AppMetricsService\",\"metadata\":{\"annotations\":{},\"name\":\"example-appmetricsservice\",\"namespace\":\"app-metrics\"},\"spec\":null}\n"}, "generation":1, "uid":"1789007d-9e64-11e9-b82f-5254001bc413", "selfLink":"", "clusterName":"", "name":"example-appmetricsservice", "namespace":"app-metrics", "creationTimestamp":"2019-07-04T14:00:34Z"}}: validation failure list:
spec in body must be of type object: "null"
```

File: pkg/controller/appmetricsservice/postgresql.go (hunk 2 above)
Comment (aliok) [reply pkg/controller/appmetricsservice/postgresql.go:18]: Let's just keep it the same with UPS. If SRE people thinks it should be configurable OR if we have time in the future, we can do it.

But it is the ideal situation.

**Review by aliok — DISMISSED** [review]
@grdryn 

The operator had some RBAC issues when I ran it inside the cluster. 

Changing the role.yaml to this fixed the problem:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: null
  name: app-metrics-operator
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - services
  - endpoints
  - persistentvolumeclaims
  - events
  - configmaps
  - secrets
  - serviceaccounts
  verbs:
  - '*'
- apiGroups:
  - route.openshift.io
  resources:
  - routes
  verbs:
  - '*'
- apiGroups:
  - apps
  resources:
  - deployments
  - daemonsets
  - replicasets
  - statefulsets
  verbs:
  - '*'
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  verbs:
  - get
  - create
- apiGroups:
  - apps
  resourceNames:
  - app-metrics-operator
  resources:
  - deployments/finalizers
  verbs:
  - update
- apiGroups:
  - metrics.aerogear.com
  resources:
  - '*'
  verbs:
  - '*'
- apiGroups:
  - image.openshift.io
  resources:
  - imagestreams
  verbs:
  - '*'
- apiGroups:
  - apps.openshift.io
  resources:
  - deploymentconfigs
  verbs:
  - '*'
```

**Comment by grdryn** [comment]
Thanks for the feedback @aliok I thought I had deployed the image to a cluster once, but clearly I had only run it locally. I'll make the changes that you've suggested, and make sure they work, and then update here.

File: deploy/crds/metrics_v1alpha1_appmetricsservice_cr.yaml (hunk 3 above)
Comment (grdryn) [reply deploy/crds/metrics_v1alpha1_appmetricsservice_cr.yaml]: Done :+1:

**Comment by grdryn** [comment]
It turns out I need to remove the oauth-proxy from here altogether (I just copied it from UPS, but it's not needed).

**Comment by grdryn** [comment]
I've removed the oauth-proxy stuff, so this should be reviewable again (and a bit smaller).

File: cmd/manager/main.go [hunk 4]
```diff
@@ -107,14 +110,32 @@ func main() {
...
+
 	// Setup all Controllers
 	if err := controller.AddToManager(mgr); err != nil {
 		log.Error(err, "")
 		os.Exit(1)
 	}
 
 	if err = serveCRMetrics(cfg); err != nil {
-		log.Info("Could not generate and serve custom resource metrics: ", err.Error())
+		log.Info("Could not generate and serve custom resource metrics", "error:", err.Error())
```
Comment (grdryn) [inline cmd/manager/main.go]: Changed this to fix an error: https://github.com/operator-framework/operator-sdk/pull/1647

File: pkg/controller/appmetricsservice/util.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	metricsv1alpha1 "github.com/aerogear/app-metrics-operator/pkg/apis/metrics/v1alpha1"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func labels(cr *metricsv1alpha1.AppMetricsService, suffix string) map[string]string {
+	return map[string]string{
+		"app":     cr.Name,
```
Comment (laurafitzgerald) [inline pkg/controller/appmetricsservice/util.go]: I need a static label on the route for the app cr. Can you add one for me please?

**Review by odra — APPROVED** [review]
followed the verification steps and it worked as expected

File: cmd/manager/main.go (hunk 4 above)
Comment (aliok) [reply cmd/manager/main.go]: +1

**Comment by aliok** [comment]
LGTM. Didn't verify it as Leo did it already.

File: pkg/controller/appmetricsservice/util.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func labels(cr *metricsv1alpha1.AppMetricsService, suffix string) map[string]string {
+	return map[string]string{
+		"app":       cr.Name,
+		"service":   fmt.Sprintf("%s-%s", cr.Name, suffix),
+		"component": "aerogear-app-metrics",
```
Comment (grdryn) [inline pkg/controller/appmetricsservice/util.go:18]: @laurafitzgerald You should be able to use this new label to find the Route from your other controller, without having to know the CR name.

**Review by grdryn — COMMENTED** [review]
Thanks for the reviews folks! I'll merge this now so that the rest of the work is unlocked :+1:

**Pull request merged**